### ISSUE TITLE
[ENDPOINT]: Proposed design change for getListing (oauth2 security and skus field)

### ISSUE THREAD

**Issue by tjoyal** [etsy/open-api#544]
### Is there an existing issue for this?

- [X] I have searched the existing tickets

### Current Endpoint Design

TLDR: The `getListing` operation is defined as "AUTHORIZATIONS: api_key" when it is possible to pass in an oauth2 header if you want the `skus` field to be returned.

I have been digging into why `getListingsByShop` was returning me the field `skus` filled where the `getListing` was not. 

Turns out the open api generator I have been using (https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/ruby.md) is following the spec to the letter and this endpoint is flagged to not accept in an oauth2 security input. In the schema (https://www.etsy.com/openapi/generated/oas/3.0.0.json) we can see that there are no security oauth2 needs for this endpoint.

The `skus` parameter seems to be the only field affected by this in `getListing`, I haven’t audited other endpoints.

### Proposed Endpoint Design Change

I am working off my limited understanding of open-api from a consumer standpoint therefore my suggestions are somewhat limited and could contain risks.

Looking at https://goswagger.io/tutorial/composed-auth/ it seems the `getListing` node could get the following security definition:
```
security: [
  {
    "api_key": []
  },
  {
    "api_key": [],
    "oauth2": ["listing_r"]
  }
]
```

That being said, at least in the ruby generators, there seems to be an issue when oauth2 is not configured for the request to be sent with a `authorization: Bearer` header without a token which is an invalid header. As one might expect, Etsy servers return `{"error":"Invalid authorization header"}` when doing so. 

I was able to make the request work by passing in a garbage access token (`authorization: Bearer invalidtoken`) but to me this is unsafe as at some point Etsy servers might start returning errors when an invalid Oauth2 token is passed in.

This should probably be corrected in the generators directly, but still for the proposal here there is a worry for this change to create a break in existing and functioning integrations which I cannot really judge on my own.

I briefly considered interesting to break down the endpoint into two distinct endpoints(eg.: `getListing` and `getPrivateListing`), but we probably want to not have to duplicate all the entry points under the “Listing” category which would be even more confusing.

So there it is, I feel the way open-api is supposed to be used is to add a secondary entry to the security array, but I am uncertain of the downstream implications.

### Why are you proposing this change?

It is unexpected, at least to my eyes, that is an accepted and undocumented security input that can be passed to the endpoint.

Arguably the `skus` field mentions “SKUs will only appear if the requesting user owns the shop.” which makes it somewhat obvious what is happening, but this warning is for humans and it cannot be processed by the ecosystem of tools built around open-api.

**Comment by etsyachristensen** [comment]
Thanks for this @tjoyal. There's a couple issues here.

First of all, our documentation OAS file is generated automatically based on the code. Unfortunately, the way our code is put together, we can't define two different expectations for security, so it will use the one as the base. Just know that if you HAVE an auth token you should always include it, not only due to this issue of SKU not returning (which will also be the case in the `getListingInventory` endpoint and the find endpoints) but also because if you use the auth token, it will not count against your global api rate limit, but instead against that shop owner's api rate limit for your app.

Secondly, we have had to diverge in a few ways such as this from the letter of the law when it comes to OpenAPI specifications. We try to adhere as much as possible but sometimes we have to make a few sacrifices.

In this particular case, we can't really change anything with the exception of possibly adding a note to the sku return field description that it will only be visible to the owner of the shop. We will look into adding that.
